permissions/resource_rule_set.go

```diff
@@ -0,0 +1,177 @@
+package permissions
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/databricks/databricks-sdk-go/service/iam"
+	"github.com/databricks/terraform-provider-databricks/common"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type RuleSetApi struct {
+	client *common.DatabricksClient
+	ctx    context.Context
+}
+
+func NewRuleSetApi(ctx context.Context, m interface{}) RuleSetApi {
+	return RuleSetApi{m.(*common.DatabricksClient), ctx}
+}
+
+func (a RuleSetApi) Read(getRuleSetReq iam.GetRuleSetRequest) (*iam.RuleSetResponse, error) {
+	if a.client.Config.AccountID != "" {
+		accountClient, err := a.client.AccountClient()
+		if err != nil {
+			return nil, err
+		}
+		return accountClient.AccessControl.GetRuleSet(a.ctx, getRuleSetReq)
+	}
+	workspaceClient, err := a.client.WorkspaceClient()
+	if err != nil {
+		return nil, err
+	}
+	return workspaceClient.AccountAccessControlProxy.GetRuleSet(a.ctx, getRuleSetReq)
+}
+
+func (a RuleSetApi) Update(updateRuleSetReq iam.UpdateRuleSetRequest) (*iam.RuleSetResponse, error) {
+	if a.client.Config.AccountID != "" {
+		accountClient, err := a.client.AccountClient()
+		if err != nil {
+			return nil, err
+		}
+		return accountClient.AccessControl.UpdateRuleSet(a.ctx, updateRuleSetReq)
+	}
+	workspaceClient, err := a.client.WorkspaceClient()
+	if err != nil {
+		return nil, err
+	}
+	return workspaceClient.AccountAccessControlProxy.UpdateRuleSet(a.ctx, updateRuleSetReq)
+}
+
+func ResourceRuleSet() *schema.Resource {
+	s := common.StructToSchema(
+		iam.RuleSetUpdateRequest{},
+		func(m map[string]*schema.Schema) map[string]*schema.Schema {
+			m["name"].Required = true
+			m["etag"] = &schema.Schema{
+				Computed: true,
+				Type:     schema.TypeString,
+			}
+			m["grant_rules"].Type = schema.TypeSet
+			common.MustSchemaPath(m, "grant_rules", "principals").Type = schema.TypeSet
+
+			return m
+		})
+	getAndUpdateRuleSet := func(ruleSetUpdateReq iam.UpdateRuleSetRequest, rsApi RuleSetApi) (string, error) {
+		ruleSetGetRes, err := rsApi.Read(iam.GetRuleSetRequest{
+			Name: ruleSetUpdateReq.Name,
+			Etag: "",
+		})
+		if err != nil {
+			return "", err
+		}
+		ruleSetUpdateReq.RuleSet.Etag = ruleSetGetRes.Etag
+		ruleSetUpdateRes, err := rsApi.Update(ruleSetUpdateReq)
+		if err != nil {
+			return "", err
+		}
+		return ruleSetUpdateRes.Etag, nil
+	}
+	handleConflictAndUpdate := func(ruleSetUpdateReq iam.UpdateRuleSetRequest, rsApi RuleSetApi) (string, error) {
+		ruleSetUpdateRes, err := rsApi.Update(ruleSetUpdateReq)
+		if err != nil {
```

Comment:
please typecast to an error and check on error code, because using exception message for logic is very fragile. See example on how to use the error code: https://github.com/databricks/cli/blob/35dd56c716eaa46c543d96a21d802614b7caf085/libs/filer/dbfs_client.go#L117-L131